### PR TITLE
Rollback env vars changed by a proc macro

### LINKED ISSUE #11355: Env vars set by proc macros should be cleared after a macro expansion
Some macros change their environment variables. For example, `sqlx` loads [all variables from `.env` file](https://github.com/launchbadge/sqlx/blob/2182925e92bc58ec4579c5b1e63377dcf5c6ae37/sqlx-macros/src/query/mod.rs#L85). These envs can affect the expansion of other macros. For example, an `sqlx` macro from crate `A` can affect `sqlx` macros from crate `B`.

Demo:

```rust
#[proc_macro]
pub fn set_var(_: TokenStream) -> TokenStream {
    std::env::set_var("FOO", "bad");
    TokenStream::new()
}

#[proc_macro]
pub fn read_var(_: TokenStream) -> TokenStream {
    let var = std::env::var("FOO").unwrap_or_else(|_| "good".to_string());
    format!("\"{}\"", var).parse().unwrap()
}
```

then use them like this:

```rust
fn main() {
    my_proc_macro::set_var!();
    let b = my_proc_macro::read_var!();
}
```

`read_var!()` should be expanded to "good", but it will likely be expanded to "bad".

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #11355

### CODE REVIEW

File: crates/proc_macro_srv/src/lib.rs [hunk 1]
```diff
@@ -101,6 +98,32 @@ impl ProcMacroSrv {
...
+    vars: Vec<(OsString, OsString)>,
+}
+
+impl EnvSnapshot {
+    fn new() -> EnvSnapshot {
+        EnvSnapshot { vars: env::vars_os().collect() }
+    }
+
+    fn rollback(self) {
+        let mut old_vars = self.vars.into_iter().collect::<HashMap<OsString, OsString>>();
```
Comment (jonas-schievink) [inline crates/proc_macro_srv/src/lib.rs:111]: Let's create the `HashMap` in `new`, to avoid the intermediate `Vec`

**Review by jonas-schievink — COMMENTED** [review]
LGTM!

bors d+

**Comment by vlad20012** [comment]
I had some discussion with @bjorn3 in issue #11355. Shouldn't we wait for his comments on this?

**Pull request merged**